src/server/pkg/sync/sync.go

```diff
@@ -0,0 +1,127 @@
+// Package sync provides utility functions similar to `git pull/push` for PFS
+package sync
+
+import (
+	"context"
+	"os"
+	"path/filepath"
+
+	pachclient "github.com/pachyderm/pachyderm/src/client"
+	"github.com/pachyderm/pachyderm/src/client/pfs"
+
+	protostream "go.pedge.io/proto/stream"
+	"golang.org/x/sync/errgroup"
+)
+
+// Pull clones an entire repo at a certain commit
+//
+// root is the local path you want to clone to
+// commit is the commit you want to clone
+// shard and diffMethod get passed to ListFile and GetFile. See documentations
+// for those functions for details on these arguments.
+func Pull(ctx context.Context, client pfs.APIClient, root string, commit *pfs.Commit, diffMethod *pfs.DiffMethod, shard *pfs.Shard) error {
+	return pullDir(ctx, client, root, commit, diffMethod, shard, "/")
+}
+
+func pullDir(ctx context.Context, client pfs.APIClient, root string, commit *pfs.Commit, diffMethod *pfs.DiffMethod, shard *pfs.Shard, dir string) error {
+	if err := os.MkdirAll(filepath.Join(root, dir), 0777); err != nil {
+		return err
+	}
+
+	fileInfos, err := client.ListFile(ctx, &pfs.ListFileRequest{
+		File: &pfs.File{
+			Commit: commit,
+			Path:   dir,
+		},
+		Shard:      shard,
+		DiffMethod: diffMethod,
+		Mode:       pfs.ListFileMode_ListFile_NORMAL,
+	})
+	if err != nil {
+		return err
+	}
+
+	var g errgroup.Group
+	for _, fileInfo := range fileInfos.FileInfo {
+		fileInfo := fileInfo
+		g.Go(func() error {
+			switch fileInfo.FileType {
+			case pfs.FileType_FILE_TYPE_REGULAR:
+				path := filepath.Join(root, fileInfo.File.Path)
+				f, err := os.Create(path)
+				if err != nil {
+					return err
+				}
+				getFileClient, err := client.GetFile(
+					ctx,
+					&pfs.GetFileRequest{
+						File: &pfs.File{
+							Commit: commit,
+							Path:   fileInfo.File.Path,
+						},
+						Shard:      shard,
+						DiffMethod: diffMethod,
+					},
+				)
+				if err != nil {
+					return err
+				}
+				if err := protostream.WriteFromStreamingBytesClient(getFileClient, f); err != nil {
+					return err
+				}
+				return f.Close()
```

Comment:
This is a file descriptor leak waiting to happen. If this function exits early then the file won't get closed. You need to `defer` the close. It should look like this:

```go
defer func() {
  if err := f.Close(); err != nil && retErr == nil {
    retErr = err
  }
}()
```